rskj-core/src/main/java/co/rsk/net/NodeBlockProcessor.java

```diff
@@ -152,224 +137,213 @@ public void processBlockHeaders(@Nonnull final MessageSender sender, @Nonnull fi
-
-        logger.trace("Missing block " + hash.toString().substring(0, 10));
-
-        sender.sendMessage(new GetBlockMessage(hash.getData()));
-
-        return;
+        Message responseMessage = new BodyResponseMessage(requestId, block.getTransactionsList(), block.getUncleList());
+        sender.sendMessage(responseMessage);
     }
 
     /**
-     * processStatus processes a Status containing another node's status (its bestBlock).
-     * If the sender has a better best block, it will be requested.
-     * Otherwise, all the blocks that the sender is missing will be sent to it.
-     *
-     * @param sender the message sender. This should be the node that sent the status message.
-     * @param status The status message containing the other node's best block.
+     * processBlockHashRequest sends the requested block body to a peer if it is available.
+     *  @param sender the sender of the BlockHashRequest message.
+     * @param requestId the id of the request
+     * @param height   the requested block's hash.
      */
     @Override
-    public void processStatus(@Nonnull final MessageSender sender, @Nonnull final Status status) {
-        logger.trace("Processing status " + status.getBestBlockNumber() + " " + Hex.toHexString(status.getBestBlockHash()).substring(0, 10) + " from " + sender.getNodeID().toString());
+    public void processBlockHashRequest(@Nonnull final MessageChannel sender, long requestId, long height) {
+        logger.trace("Processing block hash request {} {} from {}", requestId, height, sender.getPeerNodeID().toString());
+        final Block block = this.getBlockFromBlockchainStore(height);
 
-        final byte[] hash = status.getBestBlockHash();
-        nodeInformation.addBlockToNode(new ByteArrayWrapper(hash), sender.getNodeID());
-
-        if (!this.hasBlock(hash))
-            sender.sendMessage(new GetBlockMessage(hash));
+        if (block == null) {
+            // Don't waste time sending an empty response.
+            return;
+        }
 
-        final long bestBlockNumber = this.getBestBlockNumber();
-        final long peerBestBlockNumber = status.getBestBlockNumber();
+        BlockHashResponseMessage responseMessage = new BlockHashResponseMessage(requestId, block.getHash());
+        sender.sendMessage(responseMessage);
+    }
 
-        if (peerBestBlockNumber > this.lastKnownBlockNumber)
-            this.lastKnownBlockNumber = peerBestBlockNumber;
+    /**
+     * @param sender the sender of the SkeletonRequest message.
+     * @param requestId the id of the request.
+     * @param startNumber the starting block's hash to get the skeleton.
+     */
+    @Override
+    public void processSkeletonRequest(@Nonnull final MessageChannel sender, long requestId, long startNumber) {
+        logger.trace("Processing block hash request {} {} {} from {}", requestId, startNumber, sender.getPeerNodeID().toString());
+        // TODO(mc) move this to configuration
```

Comment:
If this is moved to config file we should be able to ask a peer for its value.

Reply:
It's fixed on a later branch about sync